plan/conflict.go

```diff
@@ -17,54 +17,511 @@ limitations under the License.
+			}
+		}
+		ep.Targets = []string{ep.Targets[0]}
+	case endpoint.RecordTypeCNAME:
+		for i := 1; i < len(ep.Targets); i++ {
+			if ep.Targets[i] != ep.Targets[0] {
+				return fmt.Errorf("inconsistent targets for CNAME record")
+			}
+		}
+		ep.Targets = []string{ep.Targets[0]}
+
+	default:
+		// we don't merge targets for other record types
+	}
+	return nil
+}
+
+func toEndpoint(ws wrapEndpoints) (*endpoint.Endpoint, error) {
+	if len(ws) == 0 {
+		return nil, fmt.Errorf("empty wrapEndpoints")
+	}
+	// sort by target
+	sort.Sort(ws)
+
+	ret := ws[0].endpoint.DeepCopy()
+	ret.Targets = []string{ws[0].target()}
+	for i := 1; i < len(ws); i++ {
+		wep := ws[i]
+		if ret.DNSName != wep.dnsName() {
+			return nil, fmt.Errorf("inconsistent DNSName")
+		}
+		if ret.RecordType != wep.recordType() {
+			return nil, fmt.Errorf("inconsistent RecordType")
+		}
+		if ret.SetIdentifier != wep.setIdentifier() {
+			return nil, fmt.Errorf("inconsistent SetIdentifier")
+		}
+		if ret.RecordTTL != wep.recordTTL() {
+			return nil, fmt.Errorf("inconsistent RecordTTL")
+		}
+		// check if target is duplicated
+		if ws[i-1].target() != wep.target() {
+			ret.Targets = append(ret.Targets, wep.target())
+		}
+		// src wins merge
+		ret.Labels = srcWinsMergeLabels(ret.Labels, wep.endpoint.Labels)
+		mergeProviderSpecific(ret, wep.endpoint)
+	}
+	err := mergeEndpointTargets(ret)
+
+	return ret, err
+}
+
+func (s PerResource) Resolve(currents []*endpoint.Endpoint, candidates []*endpoint.Endpoint) (changes Changes, err error) {
+	wes := make(wrapEndpoints, 0, len(currents)+len(candidates))
+	for _, ep := range currents {
+		wes, err = wes.append(ep, true)
+		if err != nil {
+			return
+		}
+	}
 	for _, ep := range candidates {
-		if ep.Labels[endpoint.ResourceLabelKey] == currentResource {
-			return ep
+		wes, err = wes.append(ep, false)
+		if err != nil {
+			return
+		}
+	}
+
+	mcac := mergeCurrentAndCandidates(wes)
+	var ep *endpoint.Endpoint
+	if onlyCurrent(mcac) {
+		// ep, err = toEndpoint(mcac)
+		// if err != nil {
+		// 	return
+		// }
+		changes.Delete = append(changes.Delete, findCurrentEndpoints(mcac)...)
+		return
+	}
+	if hasCurrentAndCandidate(mcac) {
+		currents := findCurrents(mcac)
+		sort.Sort(currents) // sort by target
+		var currentMergedLabels map[string]string
+		for _, current := range currents {
+			currentMergedLabels = srcWinsMergeLabels(currentMergedLabels, current.endpoint.Labels)
+		}
+		candidates := findCandidates(mcac)
+		sort.Sort(candidates) // sort by target
+
+		var candidateEP *endpoint.Endpoint
+		candidateEP, err = toEndpoint(candidates)
+		if err != nil {
+			return
 		}
+		candidateEP.Labels = srcWinsMergeLabels(currentMergedLabels, candidateEP.Labels)
+		oldEPs := findCurrentEndpoints(mcac)
+		foundCandidate := false
+		for _, oldEp := range oldEPs {
+			if equalEndpoint(oldEp, candidateEP) {
+				foundCandidate = true
+				continue
+			}
+			changes.UpdateOld = append(changes.UpdateOld, oldEp)
```

Comment:
Previously, `len(changes.UpdateOld) == len(changes.UpdateNew)` was an invariant. This is breaking that invariant. For what purpose?